src/test/java/org/apache/sysml/test/gpu/NeuralNetworkOpTests.java

```diff
@@ -0,0 +1,490 @@
+
+	// More comprehensive but time consuming tests
+	/*
+	private final List<Integer> Nlst = Arrays.asList(128, 64, 32);
+    private final List<Integer> Clst = Arrays.asList(30, 20, 3);
+    private final List<Integer> Hlst = Arrays.asList(400, 128, 32);
+    private final List<Integer> Wlst = Arrays.asList(400, 128, 32);
+    private final List<Integer> Klst = Arrays.asList(30, 20, 10);
+    private final List<Integer> Rlst = Arrays.asList(128, 63, 4);
+    private final List<Integer> Slst = Arrays.asList(128, 63, 4);
+    private final List<Integer> strideHeightLst = Arrays.asList(9, 3);
+    private final List<Integer> strideWidthLst = Arrays.asList(9, 3);
+    private final List<Integer> padHeightLst = Arrays.asList(3, 1);
+    private final List<Integer> padWidthLst = Arrays.asList(3, 1);
+    private final List<Double> sparsitylst = Arrays.asList(1.0);    // Only test for dense
+    */
+
+	private final List<Integer> Nlst = Arrays.asList(128, 64);
+	private final List<Integer> Clst = Arrays.asList(30, 3);
+	private final List<Integer> Hlst = Arrays.asList(256, 64);
+	private final List<Integer> Wlst = Arrays.asList(256, 64);
+	private final List<Integer> Klst = Arrays.asList(30, 20);
+	private final List<Integer> Rlst = Arrays.asList(128, 4);
+	private final List<Integer> Slst = Arrays.asList(128, 4);
+	private final List<Integer> strideHeightLst = Arrays.asList(9, 3);
+	private final List<Integer> strideWidthLst = Arrays.asList(9, 3);
+	private final List<Integer> padHeightLst = Arrays.asList(3, 1);
+	private final List<Integer> padWidthLst = Arrays.asList(3, 1);
+	private final List<Double> sparsitylst = Arrays.asList(1.0);    // Only test for dense
+
+	@Override
+	public void setUp() {
+		TestUtils.clearAssertionInformation();
+		addTestConfiguration(TEST_DIR, TEST_NAME);
+		getAndLoadTestConfiguration(TEST_NAME);
+	}
+
+	@Override
+	public double getTHRESHOLD() {
+		return 1e-5;
+	}
+
+	@Test
+	public void testConv2d() {
+		String scriptStr = "O = conv2d(image, filter, padding=[padH, padW], stride=[strideH, strideW], input_shape=[N,C,H,W], filter_shape=[K,C,R,S])";
+
+		for (long N : Nlst) {
+			for (long C : Clst) {
+				for (long H : Hlst) {
+					for (long W : Wlst) {
+						for (long K : Klst) {
+							for (long R : Rlst) {
+								for (long S : Slst) {
+									for (long strideH : strideHeightLst) {
+										for (long strideW : strideWidthLst) {
+											for (long padH : padHeightLst) {
+												for (long padW : padWidthLst) {
+													for (double sparsity : sparsitylst) {
+
+														// Make sure ops fit in GPU memory and within constraints of cudnn
+														long imageSize = N * C * H * W * 8l;
+														if (imageSize > MAX_OP_SIZE)  // image size
+															continue;
+														long filterSize = K * C * R * S * 8l;
+														if (filterSize > MAX_OP_SIZE)  // filter size
+															continue;
+														// filter is smaller than image + padding
```

Comment:
The filter could also be equal to the image+padding size.  This is used in some cases if we want to reformulate an affine layer as a convolution, which is nice if we want to do surgery on a model with an affine layer to convert it to a fully-convolutional model so that it can be applied to any image size, rather than a fixed image size as it would with an affine layer present.